lib/stdlib/src/sets.erl

```diff
@@ -483,6 +483,25 @@ filter_1(Fun, Iter) ->
             []
     end.
 
+%% map(Fun, Set) -> Set.
+%%  Map Set with Map.
+-spec map(Fun, Set1) -> Set2 when
+      Fun :: fun((Element1) -> Element2),
+      Set1 :: set(Element1),
+      Set2 :: set(Element2).
+map(F, #{}=D) when is_function(F, 1) ->
+    maps:from_keys(map_1(F, maps:iterator(D)), ?VALUE);
+map(F, #set{}=D) when is_function(F, 1) ->
+    map_set(F, D).
+
+map_1(Fun, Iter) ->
+    case maps:next(Iter) of
+        {K, _, NextIter} ->
+            [Fun(K) | map_1(Fun, NextIter)];
+        none ->
+            []
+    end.
```

Comment:
This is simpler and should also be slightly more efficient:

```suggestion
map(F, #{}=D) when is_function(F, 1) ->
    maps:from_keys([F(K) || K := _ <- D], ?VALUE);
map(F, #set{}=D) when is_function(F, 1) ->
    map_set(F, D).
```

Reply:
Oh, yeah, those will be available then =)

But wouldn't it be even simpler (dunno about efficient) to write it like this?
```erlang
map(F, #{}=D) when is_function(F, 1) ->
    #{F(K) => ?VALUE || K := _ <- D};
```

Reply:
`filter/2` could be rewritten in a similar way, like
```erlang
filter(F, #{}=D) when is_function(F, 1) ->
    #{K => ?VALUE || K := _ <- D, F(K)};
```

Reply:
Yes, simpler but less efficient. Basically, the compiler rewrites a map comprehension to a list comprehension followed by `maps:to_list/1`. For this example, it would be:

```
maps:from_list([{F(K),?VALUE} || K := _ <- D])
```

(Map generators, on the other hand, are more efficient than the corresponding calls to the iterator functions in the `maps` module because the compiler inlines most of the code for handling iterators.)

Here is the output from a run of `erlperf` showing that using `maps:from_keys/1` is faster than using a map comprehension:

```
erlperf --init_runner_all '{fun(I) -> I end, sets:from_list(lists:seq(1, 10_000), [{version,2}])}.' \
'r({F,D}) when is_function(F, 1) -> maps:from_keys([F(K) || K := _ <- D], []).' \
'r({F,D}) when is_function(F, 1) -> #{F(K) => [] || K := _ <- D}.' 
Code                                                                                  ||        QPS       Time   Rel
r({F,D}) when is_function(F, 1) -> maps:from_keys([F(K) || K := _ <- D], []).          1        916    1092 us  100%
r({F,D}) when is_function(F, 1) -> #{F(K) => [] || K := _ <- D}.                       1        817    1225 us   89%
```

Reply:
Ok, thanks for the explanation, I'll change it as suggested (and add a short comment to explain why we're using `maps:from_keys` vs a map comprehension).

Reply:
As for `filter/2`, this could still be simplified in a likewise manner, right?
```erlang
filter(F, #{}=D) when is_function(F, 1) ->
    maps:from_keys([K || K := _ <- D, F(K)], ?VALUE).
```

If you agree, should we do it in this PR while we're at it, or open a new one?

Reply:
The last commit uses a map generator as suggested.

Reply:
Yes, `filter/2` can be simplified in the same way. The simplified version is a little bit faster, especially when many elements are discarded:

```
erlperf --init_runner_all '{fun(I) -> false end, sets:from_list(lists:seq(1, 10_000), [{version,2}])}.' 'r({F,D}) -> sets:filter(F, D).' 'r({F,D}) -> t:filter(F, D).'
Code                                   ||        QPS       Time   Rel
r({F,D}) -> t:filter(F, D).             1       9393     107 us  100%
r({F,D}) -> sets:filter(F, D).          1       8441     119 us   90%
$ erlperf --init_runner_all '{fun(I) -> true end, sets:from_list(lists:seq(1, 10_000), [{version,2}])}.' 'r({F,D}) -> sets:filter(F, D).' 'r({F,D}) -> t:filter(F, D).'
Code                                   ||        QPS       Time   Rel
r({F,D}) -> sets:filter(F, D).          1        983    1017 us  100%
r({F,D}) -> t:filter(F, D).             1        947    1057 us   96%
```

The two implementations actually have similar error behaviors:

```
1> sets:filter(fun(_) -> whatever end, sets:from_list([a,b,c], [{version,2}])).
** exception error: no case clause matching whatever
     in function  sets:filter_1/2 (sets.erl, line 476)
     in call from sets:filter/2 (sets.erl, line 469)
2> t:filter(fun(_) -> whatever end, sets:from_list([a,b,c], [{version,2}])).   
** exception error: bad filter whatever
     in function  t:'-filter/2-lc$^0/1-0-'/2 (t.erl, line 6)
     in call from t:filter/2 (t.erl, line 6)
```

(For historical reasons, filters in comprehensions fail in different ways depending on whether they are guard expressions or non-guard expressions. This will be [documented in OTP 26](https://github.com/erlang/otp/pull/6929)).

We prefer a separate pull request for the update of `filter/2`.

Reply:
Yes, but you can squash the commits.